### ISSUE TITLE
some suggestions about Text Hooker

### ISSUE THREAD

**Issue by raousoft** [SourMesen/Mesen#423]
Hi dear Sour , thanks for the emulator.

We talked about the functionality of the Text Hooker here
https://github.com/SourMesen/Mesen/issues/416

I think that I would like to suggest the following steps :

- Adding Translation-ready option ( to get translation of what do we hook from google or something similar).

- Adding  FBCR like technique , to translate text directly involving OCR , like neko did in here:
http://neko.works/index.php?static=fbcr

- Adding the ability to open more than one version of the rom , and again about translation something like this person did :
https://legendsoflocalization.com/wp-content/uploads/2018/06/kefka.png

( something to do with SNES emulator , I really don't understand the concept , but maybe you will do )

This is all for now , and I will post another issue soon about Name Table , hopefully to get a help about how to fix the error , but not until I fully test it.

**Comment by wumd-ar** [comment]
Hi and thanks a lot for your efforts on the emulator. It's really exciting to see accurate NES emulation combined with debugger functionality as in-depth as no$gba or bsnes-plus.

Me and @raousoft discussed in private in great lengths about the text hooker feature and the potential therein to have "soft subtitle" translations finally for FDS games, which is a really interesting prospect. If you may accept our excited feedback, we may have some ideas for improving it.

If I understand it well, the text hooker has a database of character font tiles (graphics) associated with their match (character), which it then uses as a pseudo-OCR to compare dialogue graphics onscreen with those tiles and generate the text equivalent in a separate window. 

To my understanding, this is how FBCR generates its text from games too. The emulator fork used by Legends of Localization (and similarly, some VN hooks that detect Shift-JIS text in the application RAM) tries instead to reinterpret text data (before its graphical display is generated) and work with that, which is a much more involved (and game-specific) method.

The fatal flaw of FBCR is working from the post processed emulator image, so they request users to work with a x1, 8:7 image which isn't the best from a user perspective. This flaw is negated in your emulator's similar solution by simply applying the treatment on the image before it's post processed.

At any rate, here's my suggestion, in case you see some worth in the concept above:

- **Current Mesen support** for characters is hardcoded, and restricted to 8x8 tiles. Dakuten either omitted (can work if it's baked into the character in the font), as a separate character (か゛んは゛れ), or above the character. 
FBCR on the other hand has characters of arbitrary width and height -- which would work out greatly in case this implementation is reused for other systems like SNES etc... but I perfectly see why you would be fine the way it is right now for NES. 
**Suggestion:** In case arbitrary width is out, having presets for vertical text (with dakuten either on the right or below each letter, and indeed games with vertical text like shin onigashima are prime candidates for the soft sub treatment considering how complex they would be to romhack traditionally...) and text in 8x16 (Moon Crystal), 16x16, or mixes of 8x8 + dakuten for kana and 8x16 for kanji, would at least cover the most frequent needed use cases.

- Current Mesen just generates text data from the onscreen text.
Suggestion: The emulator generates the JP text onscreen, compares it against a translation script (xml, txt, whatever) in UTF-8 format with JP lines and replacement English lines, recognizes which line is read that moment, displays the English translation back onscreen.
Potential Issues: 
a- Some generic lines might occur more than once in the entire game script. 
Proposed solution: Use of a conditional breakpoint (if they're implemented in a similar way than FCEUX) that checks if some RAM addresses take a specific value (so we can tell who speaks, what BGM is playing, etc...) that helps differentiating that scene from others. That condition may be included in the xml for that line. I saw in some games that there's a counter of sorts telling it how far along it is reading from the actual text data in the ROM, which could be leveraged as a condition as well (in a similar way as the Legends of Localization solution).
b- Significant search time for each string. 
Proposed solution: Categorizing lines in blocks (so lines from the same scene share the same block), maybe an automatically generated index in the xml file in alphabetic order with the first few letters of each line, and their line ID for quick reference.

- Current Mesen way to display the text is on the right natively generated in Windows/Linux with modern system fonts. 
Suggestion: Offering a way (command in that xml file?) to optionally blank character font tiles with blank tiles using a similar post processing technique as the NES HD feature (depends on game, has to be provided by translation xml), and then Mesen generates a graphical render of the translation at a specific XY coordinate on top of the game's render, using a custom font (may be pixel font, vwf pixel font, or system font). A similar idea to the translated Metal Gear MSX ports on the Subsistence PS2 port.

Thanks a lot for reading this, and hopefully some of the ideas are of use.

**Comment by SourMesen** [comment]
Thanks for the feedback.  Trying to answer most points, so here goes:

-The text hooker isn't quite finished yet - I still have a couple of things I want to improve, including adding the ability to automatically translate the text via Google or some other API.

-I had never heard of FBCR until now, but yes, it's essentially the same concept (although the implementation in Mesen is far more simplistic, and probably easier/less restrictive to use, as far as NES games go).

-Dakuten are actually supported - you need to type "daku" or "han" (or the unicode characters for dakuten/handakuten: ﾞ  and ﾟ) in the corresponding tile, and then select the "Dakuten Mode" that you want to use (above, or to the right).  This still needs to be refined, I'm pretty sure I can remove the "dakuten mode" option entirely and have it automatically figure out whether it should use the dakuten on the top or the right (some games mix both in the same screen)

-Supporting characters bigger than 8x8 would require a lot of changes, and make mapping characters to tiles a lot more complex UI-wise (both for me, and for the end users). Since the majority of games use a 8x8 font, I'd prefer to stick to 8x8 tiles.

-A file with translations for the text is not a bad idea, but I think it might be simpler to just hack the ROM in many scenarios if you want it to look "nice".  It might be fairly easy to just overlay text on top of the screen (e.g like subtitles in a TV show), though, but trying to support custom fonts can be tricky when trying to make this work on multiple platforms.

In the end though, these kinds of translations would be unusable on an actual NES console and Mesen can't run on some devices at the moment (OS X, most phones), so it's not ideal.  Translating the ROM itself on the other hand will work on any decent NES emulator.

Arguably, I think it might be better to add tools that help simplify ROM hacking, if at all possible, rather than create another Mesen-only format.  (Like you mentioned, HD Packs are similar, but HD packs enable some things that no amount of rom hacking can do, whereas translating a game via rom hacking should be possible in the majority of cases?).

Something very basic, using Mesen's built-in OSD font and a simple file containing the original text + the translation (and nothing else) might be something I'm willing to add support for, but I feel like anything more complex that requires more time/effort than just translating the text would be better spent hacking the original ROM, rather than building a Mesen-only translation "patch"...
